bd-api/src/api.rs

```diff
@@ -377,9 +405,103 @@ impl Api {
       pipeline.try_load_persisted_config().await;
     }
 
+    // To make the client kill process as simple as possible we won't watch for runtime updates
+    // during operation. We will assume that runtime gets cached during the update process, and
+    // the next time the client starts it will have the new configuration. We can consider
+    // adding additional logic to handle this in the future. We can also send a synthetic
+    // unauthenticated response to the client to trigger a kill if we need to.
+    self.maybe_kill_client().await;
+
     self.maintain_active_stream().await
   }
 
+  fn kill_file_path(&self) -> PathBuf {
+    self.sdk_directory.join("client_kill_until")
+  }
+
+  fn hash_api_key(&self) -> Vec<u8> {
+    // This hash is not guaranteed to be stable across different Rust releases, but it should be
+    // stable for the lifetime of this SDK version, which is good enough for our use case. We
+    // are primarily trying to not write the actual API key into the kill file if somehow it is
+    // delivered some other way.
+    let mut hasher = DefaultHasher::new();
+    self.api_key.hash(&mut hasher);
+    hasher.finish().to_be_bytes().to_vec()
+  }
+
+  async fn maybe_kill_client(&mut self) {
+    // First we will try to read the kill file to see if it exists and we are within the kill
+    // duration.
+    let kill_until = match self.read_kill_file().await {
+      Ok(kill_until) => kill_until,
+      Err(e) => {
+        log::warn!("failed to read kill file: {}", e);
+        None
+      },
+    };
+
+    if let Some(kill_until) = kill_until {
+      let kill_until_time = kill_until.kill_until.to_offset_date_time();
+      if kill_until_time > self.time_provider.now()
+        && self.hash_api_key() == kill_until.api_key_hash
+      {
+        log::debug!(
+          "kill file exists and is still active ({} remaining), killing client",
+          kill_until_time - self.time_provider.now()
+        );
+        log::warn!(
+          "Attention: The SDK has been force disabled due to either a previous authentication \
+           failure or a remote server configuration. Double check your API key or contact support."
+        );
+        self.client_killed = true;
+      } else {
+        // Delete the kill file if the kill duration has passed. This will allow the client to
```

Comment:
Maybe note that we also wipe the file if the API key changes?

Reply:
Will do.